File: backend/repository/StockPricesRepository.js
const RepositroyAbstract = require('./RepositoryFactory');
const Database = require('../Database');
class StockPricesRepository extends RepositroyAbstract {
    static tableName = 'stock_prices';

    static async insert(stock_price,  {transaction} = {}){
        const pool = transaction || Database.getPool();
        const query = `INSERT INTO stock_prices (stock_id, price_date, open_price, close_price, high_price, low_price, volume) 
                        VALUES ($1, $2, $3, $4, $5, $6, $7)
                        RETURNING price_id`;
        const values = [stock_price.stock_id, stock_price.timestamp, stock_price.open_price, stock_price.close_price, stock_price.high_price, stock_price.low_price, stock_price.volume];
        try{
            const result = await pool.query(query, values);
            return result.rows[0].price_id;
        }catch(error){
            throw new DatabaseError(`股票資料庫新增錯誤: ${error.message}`);
        }
    }

    static async get(filters = {},  {transaction}  = {}) {
        const pool = transaction || Database.getPool();
        let query = `SELECT * FROM ${this.tableName} WHERE 1=1`;
        let values = [];
        let paramCount = 1;

        Object.entries(filters).forEach(([key, value]) => {
            if (value === null || value === undefined || key === 'end_date' || key === 'order_by') {
                return;
            }

            if (key === 'start_date') {
                if (filters.end_date) {
                    query += ` AND price_date BETWEEN $${paramCount} AND $${paramCount + 1}`;
                    values.push(value, filters.end_date);
                    paramCount += 2;
                } else {
                    query += ` AND price_date = $${paramCount}`;
                    values.push(value);
                    paramCount++;
                }
            } else if (Array.isArray(value)) {
                const placeholders = value.map(() => `$${paramCount++}`).join(', ');
                query += ` AND ${key} IN (${placeholders})`;
                values.push(...value);
            } else {
                query += ` AND ${key} = $${paramCount}`;
                values.push(value);
                paramCount++;
            }
        });

        if (filters.order_by) {
            query += ` ORDER BY ${filters.order_by}`;
        }

        try {
            const result = await pool.query(query, values);
            return result;
        } catch (error) {
            console.error('Error getting stock prices:', error);
            throw new DatabaseError(`股票價格資料庫查詢錯誤: ${error.message}`);
        }
    }

    static async delete(id,  {transaction}  = {}){
        const pool = transaction || Database.getPool();
        const query = `DELETE FROM stock_prices WHERE price_id = $1`;
        const values = [id];
        try{
            await pool.query(query, values);
        }catch(error){
            console.error('Error deleting stock price:', error);
            throw new DatabaseError(`股票資料庫刪除錯誤: ${error.message}`);
        }
    }
}

module.exports = StockPricesRepository;

File: backend/repository/stockPricesRepository.py
from backend.database.database import Database
from backend.utils.errors import DatabaseError
from backend.repository.repositoryAbs import RepositoryAbstract

class StockPricesRepository(RepositoryAbstract):
    table_name = 'stock_prices'
    key = 'price_id'
    @classmethod
    async def insert(cls, stock_price, transaction=None):
        pool = transaction or Database.get_pool()
        query = """INSERT INTO stock_prices 
                   (stock_id, price_date, open_price, close_price, high_price, low_price, volume) 
                   VALUES ($1, $2, $3, $4, $5, $6, $7)
                   RETURNING price_id"""
        values = [
            stock_price.stock_id,
            stock_price.timestamp,
            stock_price.open_price,
            stock_price.close_price,
            stock_price.high_price,
            stock_price.low_price,
            stock_price.volume
        ]
        try:
            result = await pool.fetchrow(query, *values)
            return result['price_id']
        except Exception as error:
            print('Error adding stock price:', error)
            raise DatabaseError(f"股票資料庫新增錯誤: {str(error)}")

    @classmethod
    async def get(cls, filters=None, transaction=None):
        if filters is None:
            filters = {}
        pool = transaction or Database.get_pool()
        query = f"SELECT * FROM {cls.table_name} WHERE 1=1"
        values = []
        param_count = 1

        for key, value in filters.items():
            if value is None or key == 'end_date' or key == 'order_by':
                continue

            if key == 'start_date':
                if filters.get('end_date'):
                    query += f" AND price_date BETWEEN ${param_count} AND ${param_count + 1}"
                    values.extend([value, filters['end_date']])
                    param_count += 2
                else:
                    query += f" AND price_date = ${param_count}"
                    values.append(value)
                    param_count += 1
            elif isinstance(value, list):
                placeholders = [f"${param_count + i}" for i in range(len(value))]
                query += f" AND {key} IN ({', '.join(placeholders)})"
                values.extend(value)
                param_count += len(value)
            else:
                query += f" AND {key} = ${param_count}"
                values.append(value)
                param_count += 1

        if filters.get('order_by'):
            query += f" ORDER BY {filters['order_by']}"

        try:
            result = await pool.fetch(query, *values)
            return result
        except Exception as error:
            print('Error getting stock prices:', error)
            raise DatabaseError(f"股票價格資料庫查詢錯誤: {str(error)}")

    @classmethod
    async def delete(cls, id, transaction=None):
        pool = transaction or Database.get_pool()
        query = "DELETE FROM stock_prices WHERE price_id = $1"
        values = [id]
        try:
            await pool.execute(query, *values)
        except Exception as error:
            print('Error deleting stock price:', error)
            raise DatabaseError(f"股票資料庫刪除錯誤: {str(error)}")